docs/proposals/PROPOSAL_D.md

```diff
@@ -0,0 +1,519 @@
+  - Each edge is a relationship between two nodes (e.g., `cpu` contains -> `cpu.family`.
+- A **Compatibility Specification** is compatibility metadata that describes a container image or application
+  - Includes one or more compatibility schemas
+  - Also is validated by way of links to the compspec.json for each
+
+Importantly, there is no complex graph structure in artifact itself, but rather in the single, top level schema that defines the space of metadata attributes and relationships. The tools can use the graph (nodes and edges) as needed for their use case. This means they _are not required_ to use them and can simply view the compatibility spec (what we might refer to as an artifact) as a listing of metadata attributes. This gives power and freedom to tool developers to create tools as they see fit. Several designs are afforded:
+
+1. **Simple checking of key/value pairs**: A community might have an algorithm that is as simple as their tool checking for particular keys, and that values match for a host of interest.
+2. **Graph-based logic**: A community might warrant using advanced graph parsing to do a similar assessment. An example comes from [Archspec](https://github.com/archspec), where a single architecture (from a flat artifact or the platform, depending on what the tool chooses) can [be placed in a graph](https://github.com/archspec/archspec-json/blob/master/cpu/microarchitectures.json) and used for advanced assessment of compatibility.
+
+## Modifications
+
+The compatibility schema that defines the structure of the artifacts and relationships between attributes would need to live in a single, known location, a web accessible URL that would be maintained by the compatibility interest group, and would be versioned. The artifacts would reference the version they are using. The compatibility artifact JSON blob would live alongside images in a registry.
+
+## Compatibility Schema
+
+The compatibility schema is a single source of truth (with versions) that lives in a web accessible location and defines, for a compatibility interest group:
+
+1. metadata attributes that are valid (node)
+2. relationships between attributes (edges)
+
+All nodes must be defined in the graph, and there should be a validation tool for ensuring this. Likely OCI can provide a template repository that has this validation and structure ready to go, and a compatibility interest group simply needs to fill in their metadata attribute names and relationships.
+
+### Example
+
+Note that examples are coming from [supercontainers/compspec](https://github.com/supercontainers/compspec).
+Here is an example compatibility schema that defines a set of nodes (metadata attributes) and relationships between them (edges, or how they connect to form a graph):
+
+```json
+{
+  "graph": {
+    "id": "org.supercontainers",
+    "type": "compspec",
+    "label": "compatibilities",
+    "nodes": {
+      "mpi": {
```

Comment:
another question about the schema and compatibility artifact separation. because compatibilities are relied on the schema, so if lots of compatibility artifact have been delivered to production environment, which refer to the aleady published schema version. But unfortunately, if someday a big issue found in schema, and fixed with a new version, then all big issues(the relationships) are still used in the deliverered compatibility artifacts, if we want to fix the issue, all compatibility artifacts have to be re-released using new schema.

From this perspective, self-contained schema (customizable node relationships) and compatiblity will reduce the issue propagation, and reduce the fix cost

Reply:
That's generally how it works with software too though. It's better to have versioning than not I think. Technically speaking, if you just ignore the schema (and version) you could have a "self-contained" schema. Having the entire schema within the artifact is not reasonable from a practicality standpoint. For an example, here is just the _start_ of IOR, it doesn't even include output types yet. https://github.com/supercontainers/compspec/pull/8/commits/764520d809f2f3acc40a7f1ad6ff082cd7cefc5c

Reply:
And that's just one namespace (I/O) there would be multiple defined for one file. It's huge redundancy for very little benefit IMHO. And if there were some issue with the schema, instead of updating one place you'd still need to update the many (thousands?) of artifacts instead. We likely just need some way to patch / give directive to those using the old schema.

Reply:
@ChaoyiHuang you make a good point for why we _don't_ want to embed logic for the actual choice in the artifact, because there could be some "big issue" - the implication being that it is with the logic of the selection. That is why I advocate for an approach where the compatibility specification is just that - an artifact with information, and the way that information is used to decide on image selection (the algorithm / logic) is not hard coded there. That's the main way you'd run into some issue like you are describing. If it's just adding / removing fields that is much less likely to warrant some crisis.

Reply:
>  if we want to fix the issue, all compatibility artifacts have to be re-released using new schema.

That's also the case with any of these proposals. If there is a change to anything (field, logic, etc) that warrants a change to the artifact or image manifest. I would argue my approach is _more_ flexible to that because often you can change the schema and then have some way to say "support previous versions" and there is no need to touch the artifacts. The other proposals, if they require everything hard coded into the artifact, cannot support that.  :)

Reply:
the question is about to build the artifact compatibilities specific relationship in schema: 

>if first io.archspec is about CPU, and the second one is about mpi and GPU, how to express the compatibility requirements: 1) cpu GenuineIntel amd64 + mpi v1.1 + nvidida GPU + (nvidia infiniband or arista infiband), 2) cpu GenuineAMD amd64 + mpiv1.2 + AMD GPU + arista infiniband)

> That is also up to the tool. The relationships between things are defined by the upper level schema, if that is desired, but it doesn't have to be used

These kind of comaptibilities and/or relationship and combinations is image specific, and easy to change than standard. If   they were built into schema, the problem is what I mentioned in this question.

Reply:
We can disagree then. Thanks for your feedback.

Reply:
Your argument is also akin to saying we should put the schema for an sbom in _every_ sbom because it might change. That doesn't make sense to me.